### PR TITLE
Fix thread-safety issues in CookieSettings when NavigationThreadingOptimizations is enabled.

### PR DESCRIPTION
**Pull request by goodov**
`CookieSettings::GetCookieSettingInternal` can be called from multiple threads. Fix our additions to be also thread-safe.

`ContentSettingsManagerImpl` handles mojo calls on a worker thread when `NavigationThreadingOptimizations` feature is [enabled](https://github.com/chromium/chromium/blob/102.0.4987.1/components/content_settings/browser/content_settings_manager_impl.cc#L141-L151). This feature is disabled currently via Griffin, that's why we don't see these issues in production. I've also ran some tests to ensure all calls in this area are made from UI thread when `NavigationThreadingOptimizations` is disabled.

Resolves https://github.com/brave/internal/issues/868

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/components/content_settings/core/common/cookie_settings_base.cc [hunk 1]
```diff
@@ -107,11 +111,6 @@ bool CookieSettingsBase::ShouldUseEphemeralStorage(
   return allow_1p && !allow_3p;
 }
 
-ScopedEphemeralStorageAwareness
-CookieSettingsBase::CreateScopedEphemeralStorageAwareness() const {
-  return ScopedEphemeralStorageAwareness(&ephemeral_storage_aware_, true);
```
Comment (goodov) [inline chromium_src/components/content_settings/core/common/cookie_settings_base.cc:112]: this was used only in this file thus it's safe to remove this at all.

File: chromium_src/components/content_settings/core/common/cookie_settings_base.h [hunk 2]
```diff
@@ -54,17 +49,16 @@ struct CookieSettingWithBraveMetadata {
...
                                                                                \
  private:                                                                      \
   bool IsCookieAccessAllowedImpl(                                              \
       const GURL& url, const net::SiteForCookies& site_for_cookies,            \
       const absl::optional<url::Origin>& top_frame_origin) const;              \
                                                                                \
-  mutable bool ephemeral_storage_aware_ = false;                               \
-  mutable CookieSettingWithBraveMetadata*                                      \
-      cookie_setting_with_brave_metadata_ = nullptr;                           \
+  mutable base::ThreadLocalPointer<CookieSettingWithBraveMetadata>             \
```
Comment (goodov) [inline chromium_src/components/content_settings/core/common/cookie_settings_base.h:60]: this is the main change.

File: patches/components-content_settings-core-common-cookie_settings_base.h.patch [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ class CookieSettingsBase {
+  public:
+-  CookieSettingsBase() = default;
++  CookieSettingsBase();
+ 
+   CookieSettingsBase(const CookieSettingsBase&) = delete;
+   CookieSettingsBase& operator=(const CookieSettingsBase&) = delete;
+ 
+-  virtual ~CookieSettingsBase() = default;
++  virtual ~CookieSettingsBase();
```
Comment (mkarolin) [inline patches/components-content_settings-core-common-cookie_settings_base.h.patch:16]: Not sure I am following why these changes are needed.

Comment (mkarolin) [reply patches/components-content_settings-core-common-cookie_settings_base.h.patch:16]: Ah, but of course!
